water_neuron/src/conversion.rs

```diff
@@ -15,6 +20,87 @@ use icrc_ledger_types::icrc2::transfer_from::TransferFromArgs;
 pub const MINIMUM_DEPOSIT_AMOUNT: ICP = ICP::ONE;
 pub const MINIMUM_WITHDRAWAL_AMOUNT: ICP = ICP::from_unscaled(10);
 
+pub async fn cancel_withdrawal(
+    neuron_id: NeuronId,
+) -> Result<MergeResponse, CancelWithdrawalError> {
+    let caller = ic_cdk::caller();
+    let _guard_principal = GuardPrincipal::new(caller)
+        .map_err(|guard_error| CancelWithdrawalError::GuardError { guard_error })?;
+
+    match get_full_neuron(neuron_id.id).await {
+        Ok(result) => match result {
+            Ok(neuron) => match neuron.time_left(timestamp_nanos()) {
+                Some(time) => {
+                    if time < ONE_DAY_SECONDS * 14 {
+                        return Err(CancelWithdrawalError::TooLate);
+                    }
+                }
+                None => return Err(CancelWithdrawalError::UnknownTimeLeft),
+            },
+            Err(gov_err) => return Err(CancelWithdrawalError::GovernanceError(gov_err)),
+        },
+        Err(error) => return Err(CancelWithdrawalError::GetFullNeuronError { message: error }),
+    }
+
+    let icp_due = match read_state(|s| {
+        s.neuron_id_to_withdrawal_id
+            .get(&neuron_id)
+            .and_then(|withdrawal_id| s.withdrawal_id_to_request.get(&withdrawal_id).cloned())
+    }) {
+        Some(withdrawal_request) => {
+            if withdrawal_request.receiver != caller.into() {
+                return Err(CancelWithdrawalError::BadCaller {
+                    message: format!("Caller is not the owner."),
+                });
+            }
+            withdrawal_request.icp_due
+        }
+        None => return Err(CancelWithdrawalError::RequestNotFound),
+    };
+
+    stop_dissolvement(neuron_id)
+        .await
+        .map_err(|error_msg| CancelWithdrawalError::StopDissolvementError { message: error_msg })?;
+
```

Comment:
We discussed this problem in the morning and we added a check to ensure that the neuron is more than 2 weeks from disbursing: 

```
match get_full_neuron(neuron_id.id).await {
        Ok(result) => match result {
            Ok(neuron) => match neuron.time_left(timestamp_nanos()) {
                Some(time) => {
                    if time < ONE_DAY_SECONDS * 14 {
                        return Err(CancelWithdrawalError::TooLate);
                    }
                }
                None => return Err(CancelWithdrawalError::UnknownTimeLeft),
            },
            Err(gov_err) => return Err(CancelWithdrawalError::GovernanceError(gov_err)),
        },
        Err(error) => return Err(CancelWithdrawalError::GetFullNeuronError { message: error }),
    }
```

Reply:
I am not sure the withdrawal is necessarily out of `withdrawal_to_start_dissolving` if the `stop_dissolvement` succeeds. Because the protocol tries to dissolve neurons every hour, so I think we still have to consider the case where the `withdrawal_id` is in the set.